Align Ping-From header addition with spec

achristensen07 opened the pull request:
#### b361e822845800a055fb80051acfe90e2c8d5872
<pre>
Align Ping-From header addition with spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=247448">https://bugs.webkit.org/show_bug.cgi?id=247448</a>
rdar://98338017

Reviewed by Chris Dumez.

* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendPing):

Canonical link: <a href="https://commits.webkit.org/256463@main">https://commits.webkit.org/256463@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/187660552af0fc23edb4a27de96e3ccf1fcb76d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105334 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165640 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99739 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5090 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33767 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88137 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101167 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101417 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3744 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82369 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30797 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73626 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39501 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37194 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20369 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4465 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41258 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43010 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39621 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/PingLoader.cpp:
```diff
@@ -146,9 +146,15 @@ void PingLoader::sendPing(Frame& frame, const URL& pingURL, const URL& destinati
 
     frame.loader().updateRequestAndAddExtraFields(request, IsMainResource::No);
+
+    // https://www.w3.org/TR/2009/WD-html5-20090423/history.html#hyperlink-auditing
```
Ugh, that's an old link. Could we please use this one instead?
https://html.spec.whatwg.org/multipage/links.html#hyperlink-auditing

achristensen07 replied:
Absolutely.  The old one was strange with its referer.  Thanks

cdumez commented on Source/WebCore/loader/PingLoader.cpp:
```diff
@@ -146,9 +146,15 @@ void PingLoader::sendPing(Frame& frame, const URL& pingURL, const URL& destinati
     request.setHTTPHeaderField(HTTPHeaderName::PingTo, destinationURL.string());
-    if (!SecurityPolicy::shouldHideReferrer(pingURL, frame.loader().outgoingReferrer()))
-        request.setHTTPHeaderField(HTTPHeaderName::PingFrom, document.url().string());
+    if (document.securityOrigin().isSameOriginAs(SecurityOrigin::create(pingURL).get()))
```
Could we match the conditions in the spec more closely? I find it a lot less error prone:
- If the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the [Document](https://html.spec.whatwg.org/multipage/dom.html#document) object containing the hyperlink being audited and ping URL have the [same origin](https://html.spec.whatwg.org/multipage/browsers.html#same-origin)
- If the origins are different, but the [scheme](https://url.spec.whatwg.org/#concept-url-scheme) of the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the [Document](https://html.spec.whatwg.org/multipage/dom.html#document) containing the hyperlink being audited is not "https"
request must include a `[Ping-From](https://html.spec.whatwg.org/multipage/links.html#ping-from)` header with, as its value, the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the document containing the hyperlink, and a `[Ping-To](https://html.spec.whatwg.org/multipage/links.html#ping-to)` HTTP header with, as its value, the target URL.
- Otherwise
request must include a `[Ping-To](https://html.spec.whatwg.org/multipage/links.html#ping-to)` HTTP header with, as its value, target URL. request does not include a `[Ping-From](https://html.spec.whatwg.org/multipage/links.html#ping-from)` header.

achristensen07 replied:
I was following the logic of the old algorithm quite closely, but the new link has a new algorithm :|

cdumez commented on Source/WebCore/loader/PingLoader.cpp:
```diff
@@ -146,9 +146,15 @@ void PingLoader::sendPing(Frame& frame, const URL& pingURL, const URL& destinati
+        request.setHTTPHeaderField(HTTPHeaderName::PingFrom, document.url().strippedForUseAsReferrer());
+    else if (!document.url().protocolIs("https"_s)) {
+        request.setHTTPHeaderField(HTTPHeaderName::PingFrom, document.url().strippedForUseAsReferrer());
+        request.setHTTPReferrer(document.url().strippedForUseAsReferrer());
```
Spec seems to say `no-referrer`, yet you include a referrer here? Why is that?

achristensen07 replied:
The old link said to.  That link was great feedback.

cdumez commented on LayoutTests/http/tests/navigation/ping-attribute/secure-anchor-cross-origin-expected.txt:
```diff
@@ -0,0 +1,6 @@
+Ping sent successfully
+CONTENT_TYPE: text/ping
```
This test doesn't say pass or fail? How would one know if they regress it?

achristensen07 replied:
The lack of HTTP_PING_FROM here is the important thing.  That's how all the tests in this directory are arranged.

cdumez reviewed (changes requested):
Could you please explain in the change log what the behavior change actually is?

Commit:
Align Ping-From header addition with spec
https://bugs.webkit.org/show_bug.cgi?id=247448
rdar://98338017

Reviewed by Chris Dumez.

* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendPing):

Canonical link: https://commits.webkit.org/256463@main
